src/transforms/projectionpursuit.jl

```diff
@@ -0,0 +1,203 @@
+ProjectionPursuit(perc=.85, n=50)
+ProjectionPursuit(tol=1e-4, maxiter=250, deg=5, perc=.95, n=100)
+```
+
+"""
+
+struct ProjectionPursuit{T} <: Stateless
+  tol::T
+  maxiter::Int
+  deg::Int
+  perc::T
+  n::Int
+end
+
+ProjectionPursuit(;tol=1e-6, maxiter=100, deg=5, perc=.95, n=100) =
+  ProjectionPursuit{typeof(tol)}(tol, maxiter, deg, perc, n)
+
+isrevertible(::Type{<:ProjectionPursuit}) = true
+
+# transforms a row of random variables into a convex combination 
+# of random variables with values in [-1,1] and standard normal distribution
+function rscore(Z, α)
+  ᾱ = (1/norm(α)) .* α
+  X = Z * ᾱ
+  2 .* cdf.(Normal(), X) .- 1
+end
+
+# projection index along a given direction
+function pindex(transform, Z, α)
+  X = rscore(Z, α)
+  I = (3/2) * mean(X)^2
+  if transform.deg > 1
+    Pⱼ₋₂, Pⱼ₋₁ = ones(size(X)), X
+    for j = 2:transform.deg
+      Pⱼ₋₂, Pⱼ₋₁ = 
+        Pⱼ₋₁, (1/j) * ((2j-1) * X .* Pⱼ₋₁ - (j-1) * Pⱼ₋₂)
+      I += ((2j+1)/2) * (mean(Pⱼ₋₁))^2
+    end
+  end
+  I
+end
+
+# Find axis along which projection index is maximum
+function optimaxis(transform, Z)
+  _, q = size(Z)
+  E = 1.0 * I(q)
+  f(i) = pindex(transform, Z, E[:,i])
+  Iₘₐₓ, iₘₐₓ = findmax(f, 1:q)
+  E[:,iₘₐₓ]
+end
+
+function optimdiag(transform, Z, α₀)
+  q = length(αₒ)
+  α = α₀
+  I₀ = pindex(transform, Z, α)
+  E = 1.0 * I(q)
+  for i in 1:q
+    eᵢ = E[:,i]
+    f₊ = pindex(transform, Z, (1/√(2+2α₀[i])) * (α₀ + eᵢ))
+    f₋ = pindex(transform, Z, (1/√(2-2α₀[i])) * (α₀ - eᵢ)) * (α₀[i] != 1)
+    f = (f₊ > f₋) * f₊ + (f₊ ≤ f₋) * f₋ 
+    s = (f₊ > f₋) * 1 + (f₊ ≤ f₋) * (-1)
+    if f > I₀
+      I₀ = f
+      α = (1/√(2+2s*α₀[i])) * (α₀ + s*eᵢ)
+    end
+  end
+  α
+end
+
+function optimall(transform, Z, α₀)
+  f(α) = -pindex(transform, Z, α)
+  op = optimize(f, α₀)
+  minimizer(op)
+end
+
+function optim(transform, Z)
+  α₁ = optimaxis(transform, Z)
+  α₂ = optimdiag(transform, Z, α₁)
+  optimall(transform, Z, α₂)  
+end
+
+function rmstructure(transform, Z, α)
+  q = length(α)
+  U = qr(hcat(α, rand(q, q-1)))
+  while norm(diag(U.R)) < transform.tol
+    U = qr(hcat(α, rand(q, q-1)))
+  end
+  T = Z * U.Q
+  table = Tables.table(T)
+  newtable, colcache = apply(Quantile(), table)
+  T[:,1] = collect(newtable)[1]
+  T * U.Q', U.Q, colcache
+end
+
+# projection index of the standard multivariate Gaussian
+function gausspmatrix(transform, N, q)
+  Igauss = Matrix{Float64}(undef,transform.n, q)
+  E = 1.0 * I(q)
+  for i = 1:transform.n
+    S = randn(N, q)
+    for j = 1:q
+      Igauss[i,j] = pindex(transform, S, E[:,j])
+    end
+  end
+  Igauss
+end
+
+# Computes the pindex of each loop dataset (with possibly modified structure)
+# along q random directions and compares with the array given by gaussianpmatrix.
+function stopcriteria(transform, Z, Igauss)
+  k, q = size(Igauss)
+  E = 1.0 * I(q)
+  switch = true
+  for i = 1:q
+    Iest = pindex(transform, Z, E[:,i])
+    switch *= quantile(Igauss[1:k-1,i], transform.perc) > Iest
```

Comment:
The result of the quantile function is constant for all iterations of the main loop of the algorithm. It should be precomputed instead of called inside the stopcriteria. The stopcriteria function should take as input the result of these quantile calls instead of Igauss